crates/red_knot_python_semantic/src/types.rs

```diff
@@ -467,6 +467,16 @@ impl<'db> Type<'db> {
             {
                 true
             }
+            (Type::Tuple(self_tuple), Type::Tuple(target_tuple)) => {
```

Comment:
This raises interesting questions about how we handle union-of-tuples vs tuple-of-unions, as in, the equivalence of `tuple[int, str | int]` to `tuple[int, str] | tuple[int, int]`. I'm not sure how important it is that we understand this (I don't think existing type checkers do, but it would be cool if we did.)

We could try to normalize this one way or another in type construction, at the cost of less fidelity to original annotation forms. Or we could try to handle it here, at potentially a high complexity cost. Or we could decide not to handle it at all.

Reply:
Yeah, this is an interesting question... I'd be curious to see examples of exactly which things we _should_ infer (and _would_ infer if we understood this expansion), but don't currently

Reply:
I think it's less about inference and more about simple assignability. If a function parameter is annotated as the one type and you pass it an argument that ends up getting inferred as the other, that call should be accepted but existing type checkers won't accept it.

Reply:
Ah, makes sense